girder/models/file.py

```diff
@@ -93,19 +93,40 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type contentDisposition: str or None
         :type extraParameters: str or None
         """
+        info = {'file': file, 'startByte': offset, 'endByte': endByte}
+        events.trigger('model.file.download.request', info=info)
+
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+                    file, offset=offset, headers=headers, endByte=endByte,
+                    contentDisposition=contentDisposition,
+                    extraParameters=extraParameters)
+
+                def downloadGenerator():
+                    for data in fileDownload():
+                        yield data
+                    if endByte is None or endByte >= file['size']:
+                        info['redirect'] = False
```

Comment:
Since we already sent out `info` under the `'model.file.download.request'` event, I'm concerned about directly modifying it afterwards here. I think we should be setting / adding additional properties to a copy of `info`.

Otherwise, this could have implications for debuggability and plugins that stored a copy of the original `info` object.